src/tasks/block.rs

```diff
@@ -1,329 +1,515 @@
-        }
-        trace!(confirmed = confirmed_transactions.len(), "found confirmed transactions");
+        };
 
-        // remove already-confirmed transactions
-        for transaction in confirmed_transactions {
-            in_progress.remove_tx(&transaction);
-        }
+        // Make an AlloyDB instance from the rollup provider with that latest block number
+        let alloy_db: AlloyDB<Ethereum, RuProvider> =
+            AlloyDB::new(self.ru_provider.clone(), BlockId::from(latest));
+
+        // Wrap the AlloyDB instance in a WrapDatabaseAsync and return it.
+        // This is safe to unwrap because the main function sets the proper runtime settings.
+        //
+        // See: https://docs.rs/tokio/latest/tokio/attr.main.html
+        let wrapped_db: AlloyDatabaseProvider = WrapDatabaseAsync::new(alloy_db).unwrap();
+        Some(wrapped_db)
     }
 
-    // calculate the duration in seconds until the beginning of the next block slot.
-    fn secs_to_next_slot(&self) -> u64 {
-        let curr_timestamp: u64 = SystemTime::now().duration_since(UNIX_EPOCH).unwrap().as_secs();
-        let current_slot_time = (curr_timestamp - self.config.chain_offset) % ETHEREUM_SLOT_TIME;
-        (ETHEREUM_SLOT_TIME - current_slot_time) % ETHEREUM_SLOT_TIME
+    /// Prepares the next block environment.
+    ///
+    /// Prepares the next block environment to load into the simulator by fetching the latest block number,
+    /// assigning the correct next block number, checking the basefee, and setting the timestamp,
+    /// reward address, and gas configuration for the block environment based on builder configuration.
+    ///
+    /// # Arguments
+    ///
+    /// - finish_by: The deadline at which block simulation will end.
+    async fn next_block_env(&self, finish_by: Instant) -> Result<PecorinoBlockEnv, SimulatorError> {
+        let remaining = finish_by.duration_since(Instant::now());
+        let finish_time = SystemTime::now() + remaining;
+        let deadline: DateTime<Utc> = finish_time.into();
+        tracing::debug!(deadline = %deadline, "preparing block env");
+
+        // Fetch the latest block number and increment it by 1
+        let latest_block_number = match self.ru_provider.get_block_number().await {
+            Ok(num) => num,
+            Err(err) => {
+                tracing::error!(error = %err, "RPC error during block build");
+                return Err(SimulatorError::Rpc(Report::new(err)));
+            }
+        };
+        tracing::debug!(next_block_num = latest_block_number + 1, "preparing block env");
+
+        // Fetch the basefee from previous block to calculate gas for this block
+        let basefee = match self.get_basefee().await? {
+            Some(basefee) => basefee,
+            None => {
+                tracing::warn!("get basefee failed - RPC error likely occurred");
+                BASEFEE_DEFAULT
+            }
+        };
+        tracing::debug!(basefee = basefee, "setting basefee");
```

Comment:
nit don't need the `=` unless you're modifying something: 
`debug!(basefee, "setting basefee");`